enhancements/results-forwarding.md

```diff
@@ -0,0 +1,504 @@
+Compliance Operator will not support disabling PVC storage without a valid
+forwarding endpoint.
+
+This will require changing the **aggregator** to always forward via a gRPC API
+to a configured implementation. In this case, the **aggregator** will act as a
+gRPC client. The **resultserver** then becomes a gRPC server, subject to the
+evidence forwarding provider and will be renamed **evidence-persistor** and
+responsible for writing results.
+
+Forwarder implementations would be configured via the `ScanSetting` and they'd be called "providers".
+For backwards compatibility, we'd have defaults that would point to the current mode of operation
+of the Compliance Operator. e.g. a `pvc` provider for evidence storage.
+
+For more detail on the proposed API, see the *API Extensions* section.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters, I'd like to run
+  compliance scans on my fleet in such a way that uses the minimal amount of
+  resources.
+
+* As a Site Reliability Engineer managing multiple clusters, I'd like to have
+  one place to view the compliance stance of my whole fleet.
+
+### API Extensions
+
+Given the scope of the change, we'd upgrade the `ScanSetting` version to
+`v1alpha2`. The following sections describe the additional values needed by the
+`ScanSetting` CRD.
+
+#### `resultForwarding.provider`
+
+Required: True
+Type: string
+
+The implementation to use for forwarding. The initial implementation will
+support `grpc`.
+
+#### `resultForwarding.grpc.endpoint`
+
+Required: True
+Type: string
+
+Endpoint URL to a gRPC server for result storage. The service may be in or
+outside the cluster. One potential implementation may be
+[compserv](https://github.com/rhmdnd/compserv).
+
+#### `resultForwarding.grpc.version`
+
+Required: False
+Default: `v1`
+Type: string
+
+#### `resultForwarding.grpc.tls.dynamic.caSecretName`
+
+Required: True
+Type: string
+
+Reference to a CA certificate and key pair for TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.serverSecretName`
+
+Required: True
+Type: string
+
+Reference to a server certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.tls.dynamic.clientSecretName`
+
+Required: True if using mutual TLS
+Type: string
+
+Reference to a client certificate/keypair secret for mutual TLS.
+
+#### `resultForwarding.grpc.authentication.type`
+
+Required: False
+Default: `mtls`
+Type: string
+Choices: `mtls`, `token`
+
+The type of authentication the Compliance Operator should use to authenticate
+the forwarding endpoint.
+
+#### `resultForwarding.grpc.authentication.token`
+
+Required: True if using `token` authentication type
+Type: string
+
+#### `resultForwarding.grpc.authentication.token.tokenSecretName`
+
+Required: False
+Type: string
+
+Reference to the secret containing the token.
+
+#### `resultForwarding.grpc.extraMetadata.clusterName`
+
+Required: True
+Type: string
+
+The name of the cluster to use in result payloads.
```

Comment:
AFAIK, there isn't a way in vanilla kubernetes to get this information; instead you need to specify it somehow. I believe OpenShift has the same issue. We have struggled with this with our Helm charts to the point that we have a common helm chart with these values that we set per cluster.

Reply:
Would it be possible to set it as a configuration option of the compliance operator, instead of on a per-scan-setting-binding basis?

Or is there a valid use case for having the cluster name be different across scan setting bindings?

Reply:
+1 this seems like an awkward thing to have to set up. For OCP, we could use a unique cluster ID:
```
oc get clusterversion -o jsonpath='{.items[].spec.clusterID}{"\n"}'
```

Reply:
Additionally, I wonder what multi-cluster engine uses for cluster IDs?

Reply:
@jhrozek that's a nice thing to have. But it wouldn't work for any other distro that's not OCP 😞 . What about setting a globally readable configmap with the needed values and fetching it from there? I'm not sure if there are standards out there for establishing this kind of info... everybody seems to do their own thing.

Reply:
> @jhrozek that's a nice thing to have. But it wouldn't work for any other distro that's not OCP disappointed . What about setting a globally readable configmap with the needed values and fetching it from there? I'm not sure if there are standards out there for establishing this kind of info... everybody seems to do their own thing.

As long as we can make things work automatically on distros where we /can/ detect the name automatically, then I'm fine. Creating a CM by the operator through some platform-specific code and setting the value from there (and having the option for the admin to set the values themselves if they want), then sure.

Good idea.

Reply:
Ok - I'll pull this out of the per-scan configuration then and document it else where.

Who would be responsible for setting up this config map?

Reply:
Ok - sounds like we need to pull this detail out of the configuration section since it's going to be handled globally.